switch to DefferedRegister

hanetzer opened the pull request:
Missed some stuff. Some WorldBlocks have been excluded from the conversion
for now as the slimetreefeatureconfig breaks, needs a rework. Yes, I know this will
bloat it up a bit for now, intending to rework the {Base,Block{,Item}}RegistryAdapter
classes to make use of the DefferedRegister functionality, so your combo registration
of blocks, stairs, and slabs are coming back.

KnightMiner commented:
If we are going to use deferred register, I'd like to use a few of the advantages it provides. Notably, you can register both a block and an item at the same time if desired. Additionally, since I am lazy, I'd prefer the result of the register implement `IItemProvider` so you can create stacks from it directly. See [here](https://github.com/KnightMiner/Ceramics/blob/1.15/src/main/java/knightminer/ceramics/Registration.java) for some of the tricks I used in Ceramics. Notably, a method to register stairs, slabs, and walls with a building block automatically, and a method to register all variants from an enum (most common `DyeColor`). These also then translate super well to data generators.

hanetzer commented:
Agreed, mostly this is the teardown phase. Its hard to rewrite/replace the current reg
scheme while its still in use. Would also be useful to have a variant for orestuff (do the
ore, block, oreItem, blockItem, ingot, and nuggo all in one go, along with datagen recipe
hooks). As mentioned in the pr message combo regs of block/stairs/slab/etc are coming
back.

hanetzer committed:
library: registration: crib registry framework from Ceramics

Working good so far. Have some TODO: to handle but the guts are nice.
Possible candidate for a move to Mantle?

Signed-off-by: Marty E. Plummer <hanetzer@startmail.com>

hanetzer committed:
tools: items: switch to new DeferredRegister framework

Signed-off-by: Marty E. Plummer <hanetzer@startmail.com>

hanetzer committed:
gadgets: items: switch to new DeferredRegister framework

Signed-off-by: Marty E. Plummer <hanetzer@startmail.com>

hanetzer committed:
gadgets: blocks: switch to new DeferredRegister framework

Signed-off-by: Marty E. Plummer <hanetzer@startmail.com>

hanetzer committed:
world: blocks: switch to new DeferredRegister framework

Signed-off-by: Marty E. Plummer <hanetzer@startmail.com>

hanetzer committed:
decorative: blocks: switch to new DeferredRegister framework

Signed-off-by: Marty E. Plummer <hanetzer@startmail.com>

hanetzer committed:
smeltery: blocks: switch to new DeferredRegister framework

Signed-off-by: Marty E. Plummer <hanetzer@startmail.com>

hanetzer committed:
common: blocks: switch to new DeferredRegister framework

Signed-off-by: Marty E. Plummer <hanetzer@startmail.com>

hanetzer committed:
common: items: switch to new DeferredRegister framework

Also move slime balls into FoodItems.

Signed-off-by: Marty E. Plummer <hanetzer@startmail.com>

hanetzer committed:
food: items: switch to new DeferredRegister framework

Signed-off-by: Marty E. Plummer <hanetzer@startmail.com>

hanetzer committed:
tool: parts: switch to new DeferredRegister framework

Signed-off-by: Marty E. Plummer <hanetzer@startmail.com>

hanetzer committed:
world: items: switch to new DeferredRegister framework

Signed-off-by: Marty E. Plummer <hanetzer@startmail.com>

KnightMiner commented on src/main/java/slimeknights/tconstruct/common/ClientProxy.java:
```diff
@@ -24,6 +26,6 @@ public void debug() {
   @Override
   public void spawnSlimeParticle(World world, double x, double y, double z) {
-    mc.particles.addEffect(new SlimeFxParticle(world, x, y, z, new ItemStack(CommonItems.blue_slime_ball)));
+    mc.particles.addEffect(new SlimeFxParticle(world, x, y, z, new ItemStack(FoodItems.blue_slime_ball.get())));
```
The get is unneeded, the whole point of ItemObject is it implements IItemProvider, so you can make an item stack from it.

hanetzer replied:
so nix the `.get()`? I was mostly thinking in terms of consistency.

KnightMiner replied:
I think cleaner code is more important than consistency. I would like it if `IItemProvider` was used everywhere as then we would not need the get in nearly all cases.

hanetzer committed:
fluids: switch to new DeferredRegister framework

Signed-off-by: Marty E. Plummer <hanetzer@startmail.com>

KnightMiner commented on src/main/java/slimeknights/tconstruct/fluids/TinkerFluids.java:
```diff
@@ -1,53 +1,32 @@
-  public static RegistryObject<Item> purple_slime_fluid_bucket = ITEMS.register("purple_slime_bucket", () -> new BucketItem(purple_slime_fluid, new Item.Properties().containerItem(Items.BUCKET).maxStackSize(1).group(ItemGroup.MISC)));
-  public static final SlimeFluid.Properties purple_slime_fluid_properties = new SlimeFluid.Properties(purple_slime_fluid, purple_slime_fluid_flowing, FluidAttributes.builder(FluidIcons.FLUID_STILL, FluidIcons.FLUID_FLOWING).color(0xefd236ff).density(1600).viscosity(1600).temperature(370)).explosionResistance(100.0F).bucket(purple_slime_fluid_bucket).block(purple_slime_fluid_block);
+  public static final FluidObject<ForgeFlowingFluid> blue_slime_fluid = FLUIDS.register("blue_slime_fluid", FluidAttributes.builder(FluidIcons.FLUID_STILL, FluidIcons.FLUID_FLOWING).color(0xef67f0f5).density(1500).viscosity(1500).temperature(310), Material.WATER);
+  public static final FluidObject<ForgeFlowingFluid> purple_slime_fluid = FLUIDS.register("purple_slime_fluid", FluidAttributes.builder(FluidIcons.FLUID_STILL, FluidIcons.FLUID_FLOWING).color(0xefd236ff).density(1600).viscosity(1600).temperature(370), Material.WATER);
```
Personally, I think `blue_slime` and `purple_slime` are better names, the fluid is redundant when the type is fluid, as names do not clash between registries. It also means the bucket will generate as `blue_slime_bucket` and `purple_slime_bucket`.

For the renames, if we change the name as I suggested (same for purple):
* Fluids
  * `blue_slime_fluid` -> `blue_slime` and `purple_slime_fluid` -> `purple_slime`
  * `blue_slime_fluid_flowing` -> `flowing_blue_slime` and purple
* Blocks
  * `blue_slime_fluid_block` -> `blue_slime_fluid` (I suffix `_fluid` for the block form, but leave off block because its redundant)

hanetzer replied:
using 'blue_slime' causes a crash, at least with the fluid reg class you provided in the gist.

hanetzer replied:
https://gist.github.com/hanetzer/4e874113a3d365c4c1fcdc3ce19cdb2e << nothing
changed from what's in this pr aside from renaming the reg name "blue_slime_fluid"
to "blue_slime"

hanetzer replied:
Yeah. I think the gist missed the suffix `_fluid` fixed.

KnightMiner commented on src/main/java/slimeknights/tconstruct/blocks/GadgetBlocks.java:
```diff
@@ -1,78 +1,58 @@
+  // TODO: Need a combo reg for these... two blocks and one item.
+  public static final RegistryObject<TorchBlock> stone_torch = BLOCKS.register("stone_torch", () -> new TorchBlock(BlockProperties.STONE_TORCH) {});
+  public static final RegistryObject<WallTorchBlock> wall_stone_torch = BLOCKS.register("wall_stone_torch", () -> new WallTorchBlock(BlockProperties.STONE_TORCH) {});
+  public static final ItemObject<WallOrFloorItem> stone_torch_item = ITEMS.register("stone_torch", () -> new WallOrFloorItem(stone_torch.get(), wall_stone_torch.get(), new Item.Properties().group(TinkerRegistry.tabGadgets)));
```
You can use the itemblock constructor method for this:
```java
public static final RegistryObject<WallTorchBlock> wall_stone_torch = BLOCKS.register("wall_stone_torch", () -> new WallTorchBlock(BlockProperties.STONE_TORCH) {});
public static final BlockItemObject<TorchBlock> stone_torch = BLOCKS.register("stone_torch",
    () -> new TorchBlock(BlockProperties.STONE_TORCH) {},
    (block) -> new WallOrFloorItem(block, wall_stone_torch.get(), gadgetProps));
```

If more wall or floor items come up we can reconsider, but that is pretty clean for now.

hanetzer replied:
Guess that works for now, is a bit cludgy looking but eh.

KnightMiner replied:
Better than needing to create an item registry in a block class. The block item methods exist for cases where the block item is non-normal

KnightMiner commented on src/main/java/slimeknights/tconstruct/blocks/CommonBlocks.java:
```diff
@@ -1,127 +1,67 @@
-    registry.registerBlockItem(lavawood);
-    registry.registerBlockItem(firewood);
+  /* Crafting related items */
+  public static final BlockItemObject<Block> grout = BLOCKS.register("grout", () -> new Block(BlockProperties.GENERIC_SAND_BLOCK), (b) -> new BlockTooltipItem(b, generalProps));
```
Since we use `BlockItemTooltip` everywhere, I think it might be worth making that the default in the registry. Alternatively, add a method `setDefaultBlockItem(Function<Block,? extends BlockItem> mapper)` to the block registry.

Alternatively alternatively, make a variable `private static Function<Block,? extends BlockItem> defaultItemBlock`  and pass that as a parameter to all function calls.

hanetzer replied:
I mean. we don't use it literally everywhere. Not all have a tooltip.

KnightMiner replied:
Good point. So maybe name it `tooltipBlockItem` and pass it as needed.

KnightMiner commented on src/main/java/slimeknights/tconstruct/blocks/DecorativeBlocks.java:
```diff
@@ -1,137 +1,74 @@
-  private static void registerGlass(BaseRegistryAdapter<Block> registry) {
-    registry.register(new ClearGlassBlock(BlockProperties.GENERIC_GLASS_BLOCK), "clear_glass");
+  public static final BuildingBlockObject mud_bricks = BLOCKS.registerBuilding("mud_bricks", BlockProperties.MUD_BRICKS, generalProps);
+  public static final BuildingBlockObject dried_clay = BLOCKS.registerBuilding("dried_clay", BlockProperties.DRIED_CLAY, generalProps);
```
Not a fan of dried clay getting walls since its supposed to be like vanilla terracotta. Maybe we should remove stairs and slabs from it honestly, but that might be something for a later commit.

hanetzer replied:
Yeah, I had an idea to have extended BuildingBlockObject for things which have a wall
and maybe fences for wood-like blocks.

KnightMiner replied:
For now, might be best to ditch the walls and add them back later, I need to clean up a lot of the blocks in terms of which get walls and which get stairs and slabs.

KnightMiner commented on src/main/java/slimeknights/tconstruct/blocks/GadgetBlocks.java:
```diff
@@ -1,78 +1,58 @@
+    RenderTypeLookup.setRenderLayer(wooden_dropper_rail.get(), (layer) -> layer == RenderType.getCutout());
   }
-
-  private GadgetBlocks() {}
```
Empty constructor was intentional, basically saying "do not construct this". There is a lombok annotation we could use instead though.

hanetzer replied:
Oh I've seen that one. Yeah. Gotcha.

KnightMiner commented on src/main/java/slimeknights/tconstruct/blocks/SmelteryBlocks.java:
```diff
@@ -6,223 +6,126 @@
+  }
 
-  public static final SearedGlassBlock seared_glass = injected();
+  // TODO: registerBuilding does not handle BlockTooltipItem yet.
```
Pretty easy to extend, just make a method that takes `Function<Block,? extends BlockItem>` version, or use the default thing idea I mentioned above.

We might need to consider the fact that we lose TE stairs and slabs this way. I suppose its possible to create a version of the method with 2 more parameters for the stairs and slab constructor. Let me know your opinion on that.

hanetzer replied:
Honestly, I'm thinking instead of having x amount of buildingblockobjects, we'd have three EnumObjects for fullblocks, slabs, and stairs.

KnightMiner replied:
That may work. I'm still not convinced we should have slabs and stairs for all seared variants. In 1.12 when we added them, few vanilla blocks had slabs and stairs, while now they have some logic to it.

For the sake of this PR though, if you can get the names working right, I guess doing one for blocks, one for slabs, and one for stairs are the best

hanetzer replied:
Letting you merge and reorganize before I do any more.

KnightMiner commented on src/main/java/slimeknights/tconstruct/blocks/SmelteryBlocks.java:
```diff
@@ -6,223 +6,126 @@
-  }
-
   @SubscribeEvent
   static void registerBlockItems(final RegistryEvent.Register<Item> event) {
```
Change this to `FMLCommonSetupEvent` since we no longer register any items here. Really should have done so in the first place for the block list stuff.

hanetzer replied:
Ah yeah.

KnightMiner commented on src/main/java/slimeknights/tconstruct/blocks/SmelteryBlocks.java:
```diff
@@ -6,223 +6,126 @@
+    builder.add(seared_road.get());
+    builder.add(seared_tile.get());
+
+    builder.add(seared_stone.getSlab());
```
A helper method would not be bad, pass in the immutable set builder and the building block object, adds the stairs and slab to the builder.

hanetzer replied:
Agreed, but I just wanna get this merged for now, let you reorganize, then
we can do more 'cleaning'

KnightMiner commented on src/main/java/slimeknights/tconstruct/blocks/WorldBlocks.java:
```diff
@@ -22,249 +34,177 @@
+  public static final BlockItemObject<SlimeBlock> purple_slime = BLOCKS.register("purple_slime", () -> new SlimeBlock(BlockProperties.SLIME, false), (b) -> new BlockTooltipItem(b, worldProps));
+  public static final BlockItemObject<SlimeBlock> blood_slime = BLOCKS.register("blood_slime", () -> new SlimeBlock(BlockProperties.SLIME, false), (b) -> new BlockTooltipItem(b, worldProps));
+  public static final BlockItemObject<SlimeBlock> magma_slime = BLOCKS.register("magma_slime", () -> new SlimeBlock(BlockProperties.SLIME, false), (b) -> new BlockTooltipItem(b, worldProps));
+  public static final BlockItemObject<SlimeBlock> pink_slime = BLOCKS.register("pink_slime", () -> new SlimeBlock(BlockProperties.SLIME, true), (b) -> new BlockTooltipItem(b, worldProps));
```
I would use `registerEnum` with `SlimeType` here. Register all slime types at once. Makes it easier for datagen. You do not need to pass in the full slime types values set, you can just make an array with the last five types.

hanetzer replied:
We don't have enough slimetype? Perhaps I'm missing it, was looking at what I used for the
planty stuff.

KnightMiner replied:
If the plant one is not sufficient, we should make a bigger enum with all types/expland the plant one. We can do constants like they do in Direction for the smaller lists of slime types

KnightMiner commented on src/main/java/slimeknights/tconstruct/blocks/WorldBlocks.java:
```diff
@@ -22,249 +34,177 @@
+  public static final BlockItemObject<CongealedSlimeBlock> congealed_purple_slime = BLOCKS.register("congealed_purple_slime", () -> new CongealedSlimeBlock(BlockProperties.SLIME, false), (b) -> new BlockTooltipItem(b, worldProps));
+  public static final BlockItemObject<CongealedSlimeBlock> congealed_blood_slime = BLOCKS.register("congealed_blood_slime", () -> new CongealedSlimeBlock(BlockProperties.SLIME, false), (b) -> new BlockTooltipItem(b, worldProps));
+  public static final BlockItemObject<CongealedSlimeBlock> congealed_magma_slime = BLOCKS.register("congealed_magma_slime", () -> new CongealedSlimeBlock(BlockProperties.SLIME, false), (b) -> new BlockTooltipItem(b, worldProps));
+  public static final BlockItemObject<CongealedSlimeBlock> congealed_pink_slime = BLOCKS.register("congealed_pink_slime", () -> new CongealedSlimeBlock(BlockProperties.SLIME, true), (b) -> new BlockTooltipItem(b, worldProps));
```
`registerEnum` with slime type. I think `green_congealed_slime` is fine for naming, as long as we add registry migrations.

KnightMiner commented on src/main/java/slimeknights/tconstruct/blocks/WorldBlocks.java:
```diff
@@ -22,249 +34,177 @@
+  public static final EnumObject<SlimeGrassBlock.FoliageType,SlimeGrassBlock> green_slime_grass = BLOCKS.registerEnum(SlimeGrassBlock.FoliageType.values(), "green_slime_grass", (type) -> new SlimeGrassBlock(BlockProperties.SLIME_GRASS, type), worldProps);
+  public static final EnumObject<SlimeGrassBlock.FoliageType,SlimeGrassBlock> blue_slime_grass = BLOCKS.registerEnum(SlimeGrassBlock.FoliageType.values(), "blue_slime_grass", (type) -> new SlimeGrassBlock(BlockProperties.SLIME_GRASS, type), worldProps);
+  public static final EnumObject<SlimeGrassBlock.FoliageType,SlimeGrassBlock> purple_slime_grass = BLOCKS.registerEnum(SlimeGrassBlock.FoliageType.values(), "purple_slime_grass", (type) -> new SlimeGrassBlock(BlockProperties.SLIME_GRASS, type), worldProps);
+  public static final EnumObject<SlimeGrassBlock.FoliageType,SlimeGrassBlock> magma_slime_grass = BLOCKS.registerEnum(SlimeGrassBlock.FoliageType.values(), "magma_slime_grass", (type) -> new SlimeGrassBlock(BlockProperties.SLIME_GRASS, type), worldProps);
```
If you want, `registerEnumPair` should function properly for this. Same syntax, but two enum lists and the type function has two parameters.

KnightMiner commented on src/main/java/slimeknights/tconstruct/blocks/WorldBlocks.java:
```diff
@@ -22,249 +34,177 @@
+
+  public static final BlockItemObject<SlimeVineBlock> purple_slime_vine = BLOCKS.register("purple_slime_vine", () -> new SlimeVineBlock(BlockProperties.VINE, SlimeGrassBlock.FoliageType.PURPLE, SlimeVineBlock.VineStage.START), worldProps);
+  public static final BlockItemObject<SlimeVineBlock> purple_slime_vine_middle = BLOCKS.register("purple_slime_vine_middle", () -> new SlimeVineBlock(BlockProperties.VINE, SlimeGrassBlock.FoliageType.PURPLE, SlimeVineBlock.VineStage.MIDDLE), worldProps);
+  public static final BlockItemObject<SlimeVineBlock> purple_slime_vine_end = BLOCKS.register("purple_slime_vine_end", () -> new SlimeVineBlock(BlockProperties.VINE, SlimeGrassBlock.FoliageType.PURPLE, SlimeVineBlock.VineStage.END), worldProps);
```
This is probably a good case for enum block with the enum at the end. However, maybe we should instead remove all item forms except the first and handle it like `WallOrFloorItem`. This might be better to do after this PR.

KnightMiner commented on src/main/java/slimeknights/tconstruct/common/data/TConstructBlockLootTables.java:
```diff
@@ -37,152 +39,117 @@
+
+    this.registerDropSelfLootTable(CommonBlocks.lavawood.get());
+    this.registerLootTable(CommonBlocks.lavawood.getSlab(), BlockLootTables::droppingSlab);
+    this.registerDropSelfLootTable(CommonBlocks.firewood.getStairs());
```
Make a helper to auto add dropping for a `BuildingBlockObject`. Check out [Ceramics](https://github.com/KnightMiner/Ceramics/blob/1.15/src/main/java/knightminer/ceramics/datagen/BlockLootTables.java) for an example

KnightMiner commented on src/main/java/slimeknights/tconstruct/common/data/TConstructBlockTagsProvider.java:
```diff
@@ -29,25 +30,27 @@ public TConstructBlockTagsProvider(DataGenerator generatorIn) {
-    this.getBuilder(Tags.Blocks.SLIMY_LEAVES).add(WorldBlocks.blue_slime_leaves, WorldBlocks.purple_slime_leaves, WorldBlocks.orange_slime_leaves);
-    this.getBuilder(Tags.Blocks.SLIMY_SAPLINGS).add(WorldBlocks.blue_slime_sapling, WorldBlocks.orange_slime_sapling, WorldBlocks.purple_slime_sapling);
+    this.getBuilder(Tags.Blocks.SLIMY_LOGS).add(WorldBlocks.congealed_green_slime.get(), WorldBlocks.congealed_blue_slime.get(), WorldBlocks.congealed_purple_slime.get(), WorldBlocks.congealed_blood_slime.get(), WorldBlocks.congealed_magma_slime.get());
+    for (SlimeGrassBlock.FoliageType type : SlimeGrassBlock.FoliageType.values()) {
```
In Ceramics, I made [a helper](https://github.com/KnightMiner/Ceramics/blob/1.15/src/main/java/knightminer/ceramics/datagen/BlockTagProvider.java#L72-L76) to get a list from a `EnumObject`. Might be a little cleaner than the loops.

KnightMiner commented on src/main/java/slimeknights/tconstruct/common/data/TConstructRecipeProvider.java:
```diff
@@ -57,72 +59,72 @@ protected void registerRecipes(Consumer<IFinishedRecipe> consumer) {
 
   private void addCommon(Consumer<IFinishedRecipe> consumer) {
-    ShapelessRecipeBuilder.shapelessRecipe(CommonBlocks.firewood)
+    ShapelessRecipeBuilder.shapelessRecipe(CommonBlocks.firewood.get())
```
No need for `.get` here, another case of `IItemProvider`. Really its the main reason for that, less unneeded `get` calls when possible

KnightMiner commented on src/main/java/slimeknights/tconstruct/common/data/TConstructRecipeProvider.java:
```diff
@@ -57,72 +59,72 @@ protected void registerRecipes(Consumer<IFinishedRecipe> consumer) {
       .build(consumer, "tconstruct:common/firewood/firewood");
-    ShapedRecipeBuilder.shapedRecipe(CommonBlocks.firewood_slab, 6)
-      .key('#', CommonBlocks.firewood)
+    ShapedRecipeBuilder.shapedRecipe(CommonBlocks.firewood.getSlab(), 6)
```
Use a helper to add slab and stair recipes. See [my helper from Ceramics](https://github.com/KnightMiner/Ceramics/blob/1.15/src/main/java/knightminer/ceramics/datagen/RecipeProvider.java#L429-L485)

KnightMiner commented on src/main/java/slimeknights/tconstruct/common/data/TConstructRecipeProvider.java:
```diff
@@ -149,7 +151,7 @@ private void addCommon(Consumer<IFinishedRecipe> consumer) {
     ConditionalRecipe.builder()
       .addCondition(not(new PulseLoadedCondition(TinkerPulseIds.TINKER_SMELTERY_PULSE_ID)))
-      .addRecipe(CookingRecipeBuilder.smeltingRecipe(Ingredient.fromItems(Blocks.GLASS.asItem()), DecorativeBlocks.clear_glass.asItem(), 0.1F, 200)
+      .addRecipe(CookingRecipeBuilder.smeltingRecipe(Ingredient.fromItems(Blocks.GLASS.asItem()), DecorativeBlocks.clear_glass.get().asItem(), 0.1F, 200)
```
`.asItem()` is redundant here, they implement `IItemProvider`. `.get().asItem()` is extra redundant, as `BlockItemObject` already has an `asItem` method so calling the `get` is unneeded.

KnightMiner commented on src/main/java/slimeknights/tconstruct/common/data/TConstructRecipeProvider.java:
```diff
@@ -441,149 +447,149 @@ private void addSlimeRecipes(Consumer<IFinishedRecipe> consumer) {
   private void addGlassRecipes(Consumer<IFinishedRecipe> consumer) {
-    ShapedRecipeBuilder.shapedRecipe(DecorativeBlocks.black_clear_stained_glass, 8)
-      .key('#', DecorativeBlocks.clear_glass)
+    ShapedRecipeBuilder.shapedRecipe(DecorativeBlocks.clear_stained_glass.get(ClearStainedGlassBlock.GlassColor.BLACK), 8)
```
Loop these. `DyeColor` has a method to get the relevant Forge dye tag to use in your loop.

KnightMiner commented on src/main/java/slimeknights/tconstruct/items/ToolItems.java:
```diff
@@ -3,63 +3,41 @@
+  public static final ItemObject<Item> mending_moss = ITEMS.register("mending_moss", generalProps);
+  public static final ItemObject<Item> creative_modifier = ITEMS.register("creative_modifier", generalProps);
+
+  public static final ItemObject<Item> pickaxe = ITEMS.register("pickaxe", () -> new PickaxeTool(new Item.Properties().group(TinkerRegistry.tabTools), ToolDefinitions.PICKAXE));
```
Why is this not `ItemObject<PickaxeTool>`? So you do not need the cast above?

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/registration/BlockDeferredRegister.java:
```diff
@@ -0,0 +1,166 @@
+  public BuildingBlockObject registerBuilding(final String name, Block.Properties props, Item.Properties itemProps) {
+    BlockItemObject<Block> blockObj = register(name, props, itemProps);
+    return new BuildingBlockObject(blockObj,
+      register(name + "_slab", () -> new SlabBlock(props), itemProps),
```
All of these methods would take `final Function<? super B, ? extends BlockItem> item` instead of `Item.Properties` so you can add a custom `BlockItem` method. Then add a new `registerBuilding` to call this one with just properties.

KnightMiner commented on src/main/java/slimeknights/tconstruct/items/FoodItems.java:
```diff
@@ -1,96 +1,58 @@
+  public static final ItemObject<EdibleItem> purple_slime_ball = ITEMS.register("purple_slime_ball", () -> new EdibleItem(TinkerFood.PURPLE_SLIME_BALL, TinkerRegistry.tabGeneral));
+  public static final ItemObject<EdibleItem> blood_slime_ball = ITEMS.register("blood_slime_ball", () -> new EdibleItem(TinkerFood.BLOOD_SLIME_BALL, TinkerRegistry.tabGeneral));
+  public static final ItemObject<EdibleItem> magma_slime_ball = ITEMS.register("magma_slime_ball", () -> new EdibleItem(TinkerFood.MAGMA_SLIME_BALL, TinkerRegistry.tabGeneral));
+  public static final ItemObject<EdibleItem> pink_slime_ball = ITEMS.register("pink_slime_ball", () -> new EdibleItem(TinkerFood.PINK_SLIME_BALL, TinkerRegistry.tabGeneral));
```
Use `registerEnum`. Also, this should remain in common. While it is technically edible, the main purpose of this item is a craft material, eating is just a bonus quirk. All the other items in this file are gadgets features after all.

KnightMiner commented on src/main/java/slimeknights/tconstruct/items/GadgetItems.java:
```diff
@@ -17,61 +14,46 @@
+  public static final ItemObject<SlimeSlingItem> slime_sling_purple = ITEMS.register("slime_sling_purple", SlimeSlingItem::new);
+  public static final ItemObject<SlimeSlingItem> slime_sling_magma = ITEMS.register("slime_sling_magma", SlimeSlingItem::new);
+  public static final ItemObject<SlimeSlingItem> slime_sling_green = ITEMS.register("slime_sling_green", SlimeSlingItem::new);
+  public static final ItemObject<SlimeSlingItem> slime_sling_blood = ITEMS.register("slime_sling_blood", SlimeSlingItem::new);
```
`registerEnum`. Same for slime boots. Means we will have to rename (migrations), but its more consistent with vanilla.

KnightMiner commented on src/main/java/slimeknights/tconstruct/items/GadgetItems.java:
```diff
@@ -17,61 +14,46 @@
+  public static final ItemObject<FancyItemFrameItem> ardite_item_frame = ITEMS.register("ardite_item_frame", () -> new FancyItemFrameItem(((world, pos, dir) -> new FancyItemFrameEntity(world, pos, dir, FrameType.ARDITE.getId()))));
+  public static final ItemObject<FancyItemFrameItem> manyullyn_item_frame = ITEMS.register("manyullyn_item_frame", () -> new FancyItemFrameItem(((world, pos, dir) -> new FancyItemFrameEntity(world, pos, dir, FrameType.MANYULLYN.getId()))));
+  public static final ItemObject<FancyItemFrameItem> gold_item_frame = ITEMS.register("gold_item_frame", () -> new FancyItemFrameItem(((world, pos, dir) -> new FancyItemFrameEntity(world, pos, dir, FrameType.GOLD.getId()))));
+  public static final ItemObject<FancyItemFrameItem> clear_item_frame = ITEMS.register("clear_item_frame", () -> new FancyItemFrameItem(((world, pos, dir) -> new FancyItemFrameEntity(world, pos, dir, FrameType.CLEAR.getId()))));
```
`registerEnum`, will greatly simply a method in the entity to get the item from an enum

KnightMiner commented on src/main/java/slimeknights/tconstruct/shared/BlockEvents.java:
```diff
@@ -33,11 +34,11 @@ public static void onLivingJump(LivingEvent.LivingJumpEvent event) {
     Block block = state.getBlock();
 
-    if (block == WorldBlocks.congealed_green_slime || block == WorldBlocks.congealed_blue_slime || block == WorldBlocks.congealed_purple_slime || block == WorldBlocks.congealed_blood_slime || block == WorldBlocks.congealed_magma_slime) {
+    if (block == WorldBlocks.congealed_green_slime.get() || block == WorldBlocks.congealed_blue_slime.get() || block == WorldBlocks.congealed_purple_slime.get() || block == WorldBlocks.congealed_blood_slime.get() || block == WorldBlocks.congealed_magma_slime.get()) {
```
This should really just be an instanceof check by this point. Or a tag check, but Lex dislikes tags for game logic.

KnightMiner commented on src/main/java/slimeknights/tconstruct/shared/PlayerDataEvents.java:
```diff
@@ -25,7 +25,7 @@ public static void onPlayerLoggedIn(PlayerEvent.PlayerLoggedInEvent event) {
 
       if (!data.getBoolean(TAG_PLAYER_HAS_BOOK)) {
-        ItemHandlerHelper.giveItemToPlayer(event.getPlayer(), new ItemStack(CommonItems.book));
+        ItemHandlerHelper.giveItemToPlayer(event.getPlayer(), new ItemStack(CommonItems.book.get()));
```
redundant `.get`

KnightMiner commented on src/main/java/slimeknights/tconstruct/tileentity/SmelteryTileEntities.java:
```diff
@@ -24,13 +24,13 @@
-      SmelteryBlocks.seared_road_slab, SmelteryBlocks.seared_tile_slab, SmelteryBlocks.seared_stone_stairs, SmelteryBlocks.seared_cobble_stairs, SmelteryBlocks.seared_bricks_stairs, SmelteryBlocks.seared_cracked_bricks_stairs,
-      SmelteryBlocks.seared_fancy_bricks_stairs, SmelteryBlocks.seared_square_bricks_stairs, SmelteryBlocks.seared_small_bricks_stairs, SmelteryBlocks.seared_triangle_bricks_stairs, SmelteryBlocks.seared_creeper_stairs, SmelteryBlocks.seared_paver_stairs,
-      SmelteryBlocks.seared_road_stairs, SmelteryBlocks.seared_tile_stairs, SmelteryBlocks.seared_glass), null), "smeltery_component");
+    registry.register(new TileEntityType<>(SmelteryComponentTileEntity::new, Sets.newHashSet(SmelteryBlocks.seared_stone.get(), SmelteryBlocks.seared_cobble.get(), SmelteryBlocks.seared_bricks.get(),
```
Yeah..., really should get a helper to make this less of a mess. Make a set, add blocks from `BuildingBlockObject` to it.

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/registration/object/BuildingBlockObject.java:
```diff
@@ -0,0 +1,64 @@
+    super(block.block, block.item);
+    this.slab = slab;
+    this.stairs = stairs;
+    this.wall = wall;
```
I did not see any code that handled the wall, so probably better if you remove walls for now. We need to reconsider which blocks get slabs, stairs, and walls anyways to be more consistent with vanilla.

KnightMiner commented on src/main/java/slimeknights/tconstruct/world/TinkerWorld.java:
```diff
@@ -52,76 +56,89 @@
 
-  public static SlimeTreeFeatureConfig BLUE_SLIME_TREE_CONFIG = createTreeConfig(() -> WorldBlocks.congealed_green_slime.getDefaultState(), () -> WorldBlocks.blue_slime_leaves.getDefaultState(),
-    () -> Blocks.AIR.getDefaultState(), 5, 4, false, WorldBlocks.blue_slime_sapling);
+  public static SlimeTreeFeatureConfig BLUE_SLIME_TREE_CONFIG = injected();
```
Do not use injected unless its a final field marked with `@ObjectHolder`

KnightMiner commented on src/main/java/slimeknights/tconstruct/world/TinkerWorld.java:
```diff
@@ -144,20 +161,20 @@ private static void applyFeatures() {
       .withPlacement(Placement.COUNT_RANGE.configure(new CountRangeConfig(veinCount, 32, 0, 64))));
 
-    biome.addFeature(GenerationStage.Decoration.UNDERGROUND_DECORATION, Feature.ORE.withConfiguration(new OreFeatureConfig(OreFeatureConfig.FillerBlockType.NETHERRACK, WorldBlocks.cobalt_ore.getDefaultState(), 5))
+    biome.addFeature(GenerationStage.Decoration.UNDERGROUND_DECORATION, Feature.ORE.withConfiguration(new OreFeatureConfig(OreFeatureConfig.FillerBlockType.NETHERRACK, WorldBlocks.cobalt_ore.get().getDefaultState(), 5))
```
I wonder if its worth adding `getDefaultState` to `BlockItemObject` as an alias, since we do it a lot. Can then also type `someBlock::getDefaultState` to get a provider

KnightMiner commented on src/main/java/slimeknights/tconstruct/items/CommonItems.java:
```diff
@@ -1,79 +1,46 @@
+  public static final ItemObject<Item> alubrass_nugget = ITEMS.register("alubrass_nugget", commonProps);
+  public static final ItemObject<Item> alubrass_ingot = ITEMS.register("alubrass_ingot", commonProps);
+  public static final ItemObject<Item> knightslime_nugget = ITEMS.register("knightslime_nugget", commonProps);
+  public static final ItemObject<Item> knightslime_ingot = ITEMS.register("knightslime_ingot", commonProps);
```
I think a utility for "metals" here would be good, generate ingot, nugget, and block. Maybe that makes more sense in the block registry as that has access to create blocks, do not want the item registry doing that.

hanetzer replied:
Agreed, but I just wanna get this merged for now, let you reorganize, then
we can do more 'cleaning'

KnightMiner commented on src/main/java/slimeknights/tconstruct/gadgets/entity/FrameType.java:
```diff
@@ -47,19 +47,19 @@ public String getName() {
   public static Item getFrameFromType(FrameType type) {
```
If you make fancy frames an `EnumObject`, this method becomes `return GadgetItems.item_frame.get(type);`

KnightMiner commented on src/main/java/slimeknights/tconstruct/world/block/SlimeGrassBlock.java:
```diff
@@ -158,73 +139,38 @@ private static boolean canSlimeGrassSpread(BlockState state, IWorldReader worldR
 
   public BlockState getDirtState(BlockState grassState) {
-    if (grassState.getBlock() == WorldBlocks.blue_vanilla_slime_grass || grassState.getBlock() == WorldBlocks.purple_vanilla_slime_grass || grassState.getBlock() == WorldBlocks.orange_vanilla_slime_grass) {
+    if (grassState.getBlock() == WorldBlocks.vanilla_slime_grass.get(FoliageType.BLUE) || grassState.getBlock() == WorldBlocks.vanilla_slime_grass.get(FoliageType.PURPLE) || grassState.getBlock() == WorldBlocks.vanilla_slime_grass.get(FoliageType.ORANGE)) {
```
Add a contains method to `EnumObject` for this, since we check if the block is any in the enum so much.

hanetzer replied:
Yeah, its ugly as hell doing that.

KnightMiner commented on src/main/java/slimeknights/tconstruct/fluids/TinkerFluids.java:
```diff
@@ -1,53 +1,32 @@
-  public static RegistryObject<FlowingFluidBlock> purple_slime_fluid_block = BLOCKS.register("purple_slime_fluid_block", () -> new FlowingFluidBlock(purple_slime_fluid, Block.Properties.create(Material.WATER).doesNotBlockMovement().hardnessAndResistance(100.0F).noDrops()));
-  public static RegistryObject<Item> purple_slime_fluid_bucket = ITEMS.register("purple_slime_bucket", () -> new BucketItem(purple_slime_fluid, new Item.Properties().containerItem(Items.BUCKET).maxStackSize(1).group(ItemGroup.MISC)));
-  public static final SlimeFluid.Properties purple_slime_fluid_properties = new SlimeFluid.Properties(purple_slime_fluid, purple_slime_fluid_flowing, FluidAttributes.builder(FluidIcons.FLUID_STILL, FluidIcons.FLUID_FLOWING).color(0xefd236ff).density(1600).viscosity(1600).temperature(370)).explosionResistance(100.0F).bucket(purple_slime_fluid_bucket).block(purple_slime_fluid_block);
+  public static final FluidObject<ForgeFlowingFluid> blue_slime_fluid = FLUIDS.register("blue_slime", FluidAttributes.builder(FluidIcons.FLUID_STILL, FluidIcons.FLUID_FLOWING).color(0xef67f0f5).density(1500).viscosity(1500).temperature(310), Material.WATER);
```
Slime fluid should remain, use `SlimeFluid.Still::new` and `SlimeFluid.Flowing::new` as parameters after the attributes.

hanetzer replied:
Yeah. I wasn't sure how to make use of your fluid reg at the moment I wrote this. It
*does* work tho.

KnightMiner replied:
The base Forge fluid "works", but if you check the slime fluid class, it has some extra logic in there (I think related to slime spawning)

KnightMiner commented on src/main/java/slimeknights/tconstruct/gadgets/GadgetClientEvents.java:
```diff
@@ -29,8 +29,8 @@
-    colors.register((@Nonnull ItemStack stack, int tintIndex) -> SlimeSlingItem.getColorFromStack(stack), GadgetItems.slime_sling_blue, GadgetItems.slime_sling_purple, GadgetItems.slime_sling_magma, GadgetItems.slime_sling_green, GadgetItems.slime_sling_blood);
-    colors.register((@Nonnull ItemStack stack, int tintIndex) -> SlimeBootsItem.getColorFromStack(stack), GadgetItems.slime_boots_blue, GadgetItems.slime_boots_purple, GadgetItems.slime_boots_magma, GadgetItems.slime_boots_green, GadgetItems.slime_boots_blood);
+    colors.register((@Nonnull ItemStack stack, int tintIndex) -> SlimeSlingItem.getColorFromStack(stack), GadgetItems.slime_sling_blue.get(), GadgetItems.slime_sling_purple.get(), GadgetItems.slime_sling_magma.get(), GadgetItems.slime_sling_green.get(), GadgetItems.slime_sling_blood.get());
+    colors.register((@Nonnull ItemStack stack, int tintIndex) -> SlimeBootsItem.getColorFromStack(stack), GadgetItems.slime_boots_blue.get(), GadgetItems.slime_boots_purple.get(), GadgetItems.slime_boots_magma.get(), GadgetItems.slime_boots_green.get(), GadgetItems.slime_boots_blood.get());
```
This method is so much of a mess, and its calling a messy function. To clean it up, use `registerEnum` on slime sling and slime boots, then change the code to teh following:
```java
for (SlimeType slime : SlimeType.values()) {
  colors.register((stack,index)->slime.getBallColor(), GadgetItems.slime_boots.get(slime), GadgetItems.slime_sling.get(slime));
}
```
And you can then delete the messy methods from slime sling and slime boots

hanetzer replied:
Nifty. Will do so.

KnightMiner reviewed (changes requested):
Overall, good progress. My biggest comments are about not making good use of deferred registers, as if we are going to register multiple items at once and put them in a fancy object, we should take advantage of that fancy object to write less code, not just call get on it a bunch of times.

I do see a need for data generators to run again, as you fixed some tags and changed a few names. In addition, for the changed names we need temporary migrations to allow ForgeCraft to update (can be deleted before the final release)

In a few places, I recommend writing a utility to make something cleaner. Ping me on discord if you want me to write any of those as I wrote the register wrappers. The utilities it includes right now were just what I needed in Ceramics, so of course a larger mod like this needs more utils (hence, putting it here temporarily until we move it to Mantle)

KnightMiner commented on src/main/java/slimeknights/tconstruct/blocks/GadgetBlocks.java:
```diff
@@ -1,78 +1,59 @@
-    registry.registerBlockItem(GadgetBlocks.wooden_dropper_rail);
-  }
+  public static final BlockItemObject<LadderBlock> stone_ladder = BLOCKS.register("stone_ladder", () -> new LadderBlock(BlockProperties.STONE_LADDER) {}, gadgetProps);
+  // TODO: Need a combo reg for these... two blocks and one item.
```
I'd ditch this comment now, we can reconsider the need for a combo in the future if needed, but as it stands we have no other wall/floor items

hanetzer replied:
Ah yeah. Missed it.

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/registration/BlockDeferredRegister.java:
```diff
@@ -0,0 +1,200 @@
+    return new WoodBuildingBlockObject(blockObj,
+      register(name + "_slab", () -> new SlabBlock(props), itemProps),
+      register(name + "_stairs", () -> new StairsBlock(() -> blockObj.get().getDefaultState(), props), itemProps),
+      register(name + "fence", () -> new FenceBlock(props), itemProps)
```
Should use `_fence`.

hanetzer replied:
whoops. missed that.

KnightMiner commented on .gitignore:
```diff
@@ -55,3 +55,4 @@ options.txt
 /banned-players.json
 /ops.json
 /whitelist.json
+/src/generated/resources/.cache/cache
```
Ah, no, do not delete this. I disagree with what giga said on Discord, Lex (who makes Forge) says it does not matter if you commit this file. Overall its useful to keep track of which data files are changed in commits

hanetzer replied:
problem is this file will differ for path separators depending on what platform
produces it.
`data\forge\tags\blocks\glass\black.json` becomes
`data/forge/tags/blocks/glass/black.json` when I run, which is going to make
a lot of uneeded git noise.

KnightMiner replied:
For now, either replace all with `\` for the commit, or just commit with the other `/`, we will find a way to fix it later.

KnightMiner commented on src/main/java/slimeknights/tconstruct/TConstruct.java:
```diff
@@ -142,4 +170,37 @@ public Container createMenu(int p_createMenu_1_, PlayerInventory p_createMenu_2_
+    for (RegistryEvent.MissingMappings.Mapping<Item> entry : event.getAllMappings()) {
+      for (SlimeBlock.SlimeType slime : SlimeBlock.SlimeType.values()) {
+        // Remap slime_sling_$slime
+        if (entry.key.getNamespace().equals(TConstruct.modID) && entry.key.getPath().equals("slime_sling_" + slime.getName())) {
```
Add a single mod ID filter at the top, can save the whole effort of the loop if its not our mod

KnightMiner commented on src/main/java/slimeknights/tconstruct/TConstruct.java:
```diff
@@ -142,4 +170,37 @@ public Container createMenu(int p_createMenu_1_, PlayerInventory p_createMenu_2_
   }
+
+  @SubscribeEvent // TODO: Remove after a while, maybe at release. Doesn't work right now?
+  public void missingItemMappings(RegistryEvent.MissingMappings<Item> event) {
```
Is this still not working? Good chance its the wrong event bus, I think this is on the FORGE bus instead of the MOD one, but I may be backwards as all other reg events run on the MOD bus.

Good way to check is to add a breakpoint at the start of the function, or a LOG print, and see if it is hit.

hanetzer replied:
works. It wasn't working when I originally wrote it inside the respective
submodule I was renaming blocks/items in (eg WorldBlocks). There
was also something else being stupid but I forget what it was, but its
all good now.

KnightMiner commented on src/main/java/slimeknights/tconstruct/blocks/GadgetBlocks.java:
```diff
@@ -9,70 +10,50 @@
-
-    registry.register(new PunjiBlock(BlockProperties.PUNJI), "punji");
+  private static final BlockDeferredRegister BLOCKS = new BlockDeferredRegister(TConstruct.modID);
+  private static final ItemDeferredRegister ITEMS = new ItemDeferredRegister(TConstruct.modID);
```
Should be able to remove items here, I assume its unused now

KnightMiner commented on src/main/java/slimeknights/tconstruct/blocks/WorldBlocks.java:
```diff
@@ -22,249 +37,180 @@
 
+  private static final Item.Properties worldProps = new Item.Properties().group(TinkerRegistry.tabWorld);
+  private static final BlockDeferredRegister BLOCKS = new BlockDeferredRegister(TConstruct.modID);
+  private static final ItemDeferredRegister ITEMS = new ItemDeferredRegister(TConstruct.modID);
```
Shouldn't need the items registerer here

KnightMiner commented on src/main/java/slimeknights/tconstruct/common/data/TConstructBlockLootTables.java:
```diff
@@ -133,137 +123,103 @@ private void addTools() {
-    this.registerLootTable(WorldBlocks.purple_slime_tall_grass, BlockLootTables::onlyWithShears);
-    this.registerLootTable(WorldBlocks.orange_slime_tall_grass, BlockLootTables::onlyWithShears);
+    for (SlimeBlock.SlimeType slime : SlimeBlock.SlimeType.values()) {
+      this.registerDropSelfLootTable(WorldBlocks.slime.get(slime));
```
Skip vanilla green in this loop. Might be easier to do a separate loop, one over TINKERS slimes and one over all.

KnightMiner commented on src/main/java/slimeknights/tconstruct/common/data/TConstructRecipeProvider.java:
```diff
@@ -671,225 +570,60 @@ private void addStoneRecipes(Consumer<IFinishedRecipe> consumer) {
-        .patternLine("# #")
-        .addCriterion("has_item", this.hasItem(Items.SLIME_BALL))::build)
-      .setAdvancement(new ResourceLocation(TConstruct.modID, "recipes/tinkers_gadgets/slimeboots/blood"), ConditionalAdvancement.builder()
+    for (SlimeBlock.SlimeType slime : SlimeBlock.SlimeType.values()) {
```
So, the thing special about pink slime boots is they can be crafted from any slime type. However, I think we are missing the recipe type for that recipe (prevents conflicts with the colored boots), so maybe leave this as pink for now and I will add it in a later commit.

hanetzer replied:
True, but you should also be able to produce a pink from a solid pink
craft. Perhaps after the loopy stuff make a specific pink_from_mixed
recipe manually.

hanetzer replied:
using tags, obviously. Our and vanilla slime's are all tagged under slimeballs, right?

KnightMiner replied:
I had code in 1.12 that basically defined a list of "matches", so pink was actually handled through a fallback because pink lacked a recipe. I will look into the best way to remake it a bit, maybe just list the "unless" recipes would be cleaner

hanetzer replied:
Yeah. Well, we'll deal with that later.

KnightMiner commented on src/main/java/slimeknights/tconstruct/items/FoodItems.java:
```diff
@@ -1,96 +1,70 @@
+  /* Slime Balls are edible, believe it or not */
+  public static final EnumObject<SlimeBlock.SlimeType, Item> slime_ball;
+  static {
+    EnumObject<SlimeBlock.SlimeType,EdibleItem> tinkerSlimeballs = ITEMS.registerEnum(SlimeBlock.SlimeType.TINKER, "slime_ball", (type) -> new EdibleItem(type.getSlimeFood(type), TinkerRegistry.tabGeneral));
```
Since we do this twice, I almost wonder if we should have a more streamlined way to do it.

* One option is passing a `Consumer<Map<T,Supplier<? extends I>` as an optional parameter to `registerEnum`, which will be called to add extra data to the map before returning.
* Another option which may be a bit easier to write is adding a method that returns the map directly, either from `EnumObject` (as a copy) or the deferred register, so we can add extra items.

hanetzer replied:
Yeah, that could be helpful, but I think its outside my grok atm.

KnightMiner replied:
Yeah, just leave it and I can deal with it later if it becomes common then.

KnightMiner commented on src/main/java/slimeknights/tconstruct/items/GadgetItems.java:
```diff
@@ -17,61 +17,42 @@
+    ITEMS.register(modEventBus);
+    for (SlimeBlock.SlimeType type : SlimeBlock.SlimeType.values()) {
+      System.out.println(type);
+      System.out.println(type.toString());
```
Leftover debug?

hanetzer replied:
yep. part of why the thing wasn't working with remaping; SlimeType's toString
is not overriden for lowercase.

KnightMiner commented on src/main/java/slimeknights/tconstruct/shared/block/SlimeBlock.java:
```diff
@@ -61,6 +69,44 @@ public int getBallColor() {
+      return slimeBallTag;
+    }
+
+    public Food getSlimeFood(SlimeType type) {
```
Toss some JavaDoc headers on these methods. I like to have all methods either have a `@Override` or a javadoc. I know not all of the project follows that, but if we add java docs for new methods it will be less work when I go back and add javadocs for all that we missed.

KnightMiner commented on src/main/java/slimeknights/tconstruct/shared/block/SlimeBlock.java:
```diff
@@ -30,22 +35,25 @@ public void fillItemGroup(ItemGroup group, NonNullList<ItemStack> items) {
     private final int color, ballColor;
+    private final Tag<Item> slimeBallTag;
     public static final SlimeType[] VISIBLE_COLORS = {GREEN, BLUE, PURPLE, BLOOD, MAGMA};
+    public static final SlimeType[] TINKER = {BLUE, PURPLE, BLOOD, MAGMA, PINK};
```
Likewise, a JavaDoc on this array would not hurt. Explain what makes this different from `SlimeType.values()` and `SlimeType.VISIBLE_COLORS` (visible colors could use a javadoc too)

hanetzer replied:
well, I have no idea what VISIBLE_COLORS is being used for, idea says
its not used. And I have no idea what its supposed to be for, so can't really
document it.

KnightMiner replied:
Visible colors is probably leftover from 1.12. In 1.12, we hid pink from creative tabs as its a fallback, not a main slime color. All the other types occur naturally. Feel free to leave no comment on it though, since its currently unused.

KnightMiner commented on src/main/java/slimeknights/tconstruct/world/worldgen/islands/SlimeIslandVariant.java:
```diff
@@ -6,22 +6,24 @@
-    createArray(WorldBlocks.congealed_magma_slime.getDefaultState(), WorldBlocks.congealed_blood_slime.getDefaultState()),
-    null, createArray(WorldBlocks.orange_slime_fern.getDefaultState(), WorldBlocks.orange_slime_tall_grass.getDefaultState()));
+  BLUE(0, WorldBlocks.blue_slime_grass.get(SlimeGrassBlock.FoliageType.BLUE).getDefaultState(), TinkerFluids.blue_slime_fluid.getBlock(),
+    createArray(WorldBlocks.congealed_slime.get(SlimeBlock.SlimeType.BLUE).getDefaultState(), WorldBlocks.congealed_slime.get(SlimeBlock.SlimeType.BLUE).getDefaultState()),
```
Second one in this array should be green

KnightMiner commented on src/main/java/slimeknights/tconstruct/world/worldgen/islands/SlimeIslandVariant.java:
```diff
@@ -6,22 +6,24 @@
+    createArray(WorldBlocks.congealed_slime.get(SlimeBlock.SlimeType.BLUE).getDefaultState(), WorldBlocks.congealed_slime.get(SlimeBlock.SlimeType.BLUE).getDefaultState()),
+    WorldBlocks.blue_slime_vine.get().getDefaultState(), createArray(WorldBlocks.slime_fern.get(SlimeGrassBlock.FoliageType.PURPLE).getDefaultState(), WorldBlocks.slime_tall_grass.get(SlimeGrassBlock.FoliageType.PURPLE).getDefaultState())),
+  GREEN(1, WorldBlocks.green_slime_grass.get(SlimeGrassBlock.FoliageType.BLUE).getDefaultState(), TinkerFluids.blue_slime_fluid.getBlock(),
+    createArray(WorldBlocks.congealed_slime.get(SlimeBlock.SlimeType.BLUE).getDefaultState(), WorldBlocks.congealed_slime.get(SlimeBlock.SlimeType.BLUE).getDefaultState()),
```
Same here, second is green

KnightMiner commented on src/main/resources/assets/tconstruct/lang/en_us.json:
```diff
@@ -219,32 +219,32 @@
     "item.tconstruct.efln_ball": "EFLN",
```
Congealed slimes also had an ID change, make sure you get those in lang.

The recipe renames should help you with a list of all changed IDs
